### ISSUE TITLE
selecting and creating inside an element

### ISSUE THREAD

**Issue by shiffman** [processing/p5.js#944]
Another topic for @futuremarc and @lmccart upcoming meeting about p5.dom.   This came up with a homework assignment this week, see this discussion:

https://github.com/rlieberman/programming-a-to-z/pull/1

Should `select()` and `selectAll()` be available to operate on a `p5.Element` (looking only inside those elements)?  Or can that be done with some syntax passed into `select()` or `selectAll()`.

I had previously asked (can't find the issue) if you should be able to call the `createX()` methods on a `p5.element` and have them appended to that element rather than body, but we decided it was easy enough to use `parent()`.  I think this question is somewhat related?

**Comment by lmccart** [comment]
instead of adding `select()` and `selectAll()` as methods of the `p5.Element`, let's pass in the container to search as optional second arg (defaults to body). so, three cases to support:

1. `var elts = selectAll('#blah');`
2. `var elts = selectAll('#blah', '#cooldiv');`
3. `var elts = selectAll('#blah', myDiv);`

**Comment by DarkPrince304** [comment]
@shiffman Any specific use case for selecting one element existing inside the other?

**Comment by lmccart** [comment]
it's not super easy to handle because for performance reasons we are using the `getElementByID()` or `getElementbyClassName()` etc calls, rather than `querySelector()` which supports this out of the box. so we decided this was sort of overkill to support, if the user wants something this specific they can at this point learn about `querySelector()` or maybe even make the switch to sizzle.js or jquery if it's a better fit for their project.

**Comment by shiffman** [comment]
@lmccart my recollection from our meeting is that we decided not to support all of the `querySelector()` syntax but that we would overload `select()` to allow a second argument (`id` or `p5.Element`) to seach in.  So the use case would be:

```html
<div id = "container">
  <p>one</p>
  <p>two</p>
  <p>three</p>
</div>
<p>Another paragraph outside the container</p>
```

And then all of the following would grab only the `<p>` elements inside `container`:

```javascript
 // Option 1 (2nd arg is always an id so # assumed)
var elts = select('p', 'container'); 

// Option 2 (Also ok to use #)
var elts = select('p', '#container');  

// Option 3, pass in p5 element instance rather than id
var container = select('#container');
var elts = select('p', container);
```

I think this may be worth adding, but am also fine with skipping and pointing people to other DOM manipulation libraries.

**Comment by lmccart** [comment]
ah, yes, I misunderstood the question. the cases you list here are the same in my post above sep 29. what I meant was, we won't be supporting multi-part strings like `select('p', '#a div');

**Comment by DarkPrince304** [comment]
Lets include this in our own DOM library so that people dont have to go through the burden of choosing another library. I would like to work on this :)

**Comment by lmccart** [comment]
ok great, I just want to check first with @futuremarc that he hasn't already begun work on this, and it's good for @DarkPrince304 to go forward implementing support for the three cases outlined by @shiffman above?

**Comment by DarkPrince304** [comment]
Hello @lmccart , I have implemented all the test-cases as mentioned by @shiffman . You can check it out in my "selectArg" branch. I havn't made a PR because we don't know whether @futuremarc has started this :)

**Comment by futuremarc** [comment]
Go ahead!

**Comment by lmccart** [comment]
closed with #944
